### PR TITLE
[aptos-cli] Add ability to publish modules

### PR DESCRIPTION
**Pull request by gregnazario**
Publish modules via the CLI with a single command.  In general, after this PR I want to clean up the common parameters and move things to the config & clean up other things.

```
$ aptos move publish --package-dir /opt/git/aptos-core/aptos-move/move-examples/ --chain-id 9
{
  "Result": {
    "type": "user_transaction",
    "version": "781024",
    "hash": "0xd151922f24bb92b4a389e1f819fa2232615308279f2472de751da057f173decf",
    "state_root_hash": "0x70965acca308a860ec0aafa471987265d8307aefb60663be48233dc10ed27ce6",
    "event_root_hash": "0x414343554d554c41544f525f504c414345484f4c4445525f4841534800000000",
    "gas_used": "9",
    "success": true,
    "vm_status": "Executed successfully",
    "accumulator_root_hash": "0x792251f254accbc83bbb58e5ab45376dd3b4a37cc7ade340d421465843c73cd3",
    "sender": "0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3",
    "sequence_number": "2",
    "max_gas_amount": "1000",
    "gas_unit_price": "1",
    "gas_currency_code": "XUS",
    "expiration_timestamp_secs": "1650392416",
    "payload": {
      "type": "module_bundle_payload",
      "modules": [
        {
          "bytecode": "0xa11ceb0b040000000b01000a020a12031c2504410405452d0772e00108d202400692030a0a9c03150cb103650d96040400000101010201030104000506000006080001070700030e0401060100080001000009020300020f0404000110060100041107000003120709010603130a030106050806080105010802020c0a02000103040508020802070801010a0201060c010800010b0301090002070b030109000900074d657373616765054153434949064572726f7273054576656e74065369676e6572124d6573736167654368616e67654576656e740d4d657373616765486f6c64657206537472696e670b6765745f6d6573736167650b7365745f6d6573736167650c66726f6d5f6d6573736167650a746f5f6d657373616765076d657373616765156d6573736167655f6368616e67655f6576656e74730b4576656e7448616e646c650d6e6f745f7075626c697368656406737472696e670a616464726573735f6f66106e65775f6576656e745f68616e646c650a656d69745f6576656e74842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e30000000000000000000000000000000000000000000000000000000000000001030800000000000000000002020a08020b08020102020c08020d0b030108000001000101030b0a002901030607001102270b002b0110001402010200010105240b0111030c040e0011040c020a02290120030b05120e000b040e00380012012d0105230b022a010c050a051000140c030a050f010b030a04120038010b040b050f0015020100010100",
          "abi": null
        }
      ]
    },
    "signature": {
      "type": "ed25519_signature",
      "public_key": "0x85b7b64b9d26eb5c69357e7bfb68642e7125c364782b08d49b93a943f92e525d",
      "signature": "0xc2d78fd1484514ad2c26e8b3e32f7ef4d830d91ddd9087d3c91580bc5d9bbfdd2f438de2717b89fee511a2c6c6bb3867cdfd0ccde08346793e8045b9065c5903"
    },
    "events": [],
    "timestamp": "1650392386578052"
  }
}
```

Running also works now:
```
$ aptos move run --chain-id 9 --function-id 0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3::Message::set_message 00
{
  "Result": {
    "type": "user_transaction",
    "version": "826944",
    "hash": "0xb4d96a95d61a1893b0a4cdf5d06d09a54fcd48f3d8e2e3e8f3dfc15f5badabee",
    "state_root_hash": "0x4bdd2f6facf722036c162d9017143b547a1d8c814e93330490791d2be5c71f78",
    "event_root_hash": "0x0084ef40f2dec31ca10c6f18adb25da9020b26c555151a74396a664e1ab38f41",
    "gas_used": "68",
    "success": true,
    "vm_status": "Executed successfully",
    "accumulator_root_hash": "0x825b9e402e4c16654d535a8ff3002a231974c1dad0fb63ff90242c52bc95d12e",
    "sender": "0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3",
    "sequence_number": "45",
    "max_gas_amount": "1000",
    "gas_unit_price": "1",
    "gas_currency_code": "XUS",
    "expiration_timestamp_secs": "1650413800",
    "payload": {
      "type": "script_function_payload",
      "function": "0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3::Message::set_message",
      "type_arguments": [],
      "arguments": [
        "0x"
      ]
    },
    "signature": {
      "type": "ed25519_signature",
      "public_key": "0x85b7b64b9d26eb5c69357e7bfb68642e7125c364782b08d49b93a943f92e525d",
      "signature": "0x2964e3105b96755840daf007157b95da502110a2fab2be5a3b163f8150073bb9b4304d49c36dade672d0aca8346dd5c04f2f0d30ffe90f4c599a36425f6ce40a"
    },
    "events": [
      {
        "key": "0x0200000000000000842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3",
        "sequence_number": "11",
        "type": "0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3::Message::MessageChangeEvent",
        "data": {
          "from_message": "",
          "to_message": ""
        }
      }
    ],
    "timestamp": "1650413770266663"
  }
}

```

### CODE REVIEW

File: crates/aptos/src/move_tool/mod.rs [hunk 1]
```diff
@@ -98,3 +109,73 @@ fn compile_move(build_config: BuildConfig, package_dir: &Path) -> Result<Compile
         .compile_package(package_dir, &mut Vec::new())
         .map_err(|err| Error::MoveCompiliationError(err.to_string()))
 }
+
+/// Run Move unit tests against a package path
```
Comment (CapCap) [inline crates/aptos/src/move_tool/mod.rs:113]: this might be copypasta

Comment (gregnazario) [reply crates/aptos/src/move_tool/mod.rs:122]: Yeah, I'll fix it

File: crates/aptos/src/move_tool/mod.rs [hunk 2]
```diff
@@ -98,3 +118,173 @@ fn compile_move(build_config: BuildConfig, package_dir: &Path) -> Result<Compile
...
+                    hex::decode(stripped_hex).unwrap()
+                } else {
+                    hex::decode(str).unwrap()
+                }
+            })
+            .collect();
+
+        let script_function = ScriptFunction::new(module_id, function, Vec::new(), args);
+        println!("Encoded: {}", hex::encode("Hi Greg"));
+        println!("{:?}", serde_json::to_string(&script_function));
```
Comment (kent-white) [inline crates/aptos/src/move_tool/mod.rs:234]: leftover testing?

File: crates/aptos/src/move_tool/mod.rs [hunk 3]
```diff
@@ -98,3 +118,173 @@ fn compile_move(build_config: BuildConfig, package_dir: &Path) -> Result<Compile
...
+        // Now that it's compiled, lets send it
+        let sender_key = if let Some(private_key) = self
+            .private_key_options
+            .extract_private_key(self.encoding_options.encoding)?
+        {
+            private_key
+        } else {
+            let config = CliConfig::load()?;
+            config.private_key.unwrap()
+        };
```
Comment (kent-white) [inline crates/aptos/src/move_tool/mod.rs:244]: another reason to move it to `extract_private_key`

**Review by kent-white — APPROVED** [review]
Looks good, at least move the leftover prints

File: crates/aptos/src/move_tool/mod.rs [hunk 4]
```diff
@@ -98,3 +118,161 @@ fn compile_move(build_config: BuildConfig, package_dir: &Path) -> Result<Compile
...
+
+impl PublishPackage {
+    pub async fn execute(&self) -> Result<aptos_rest_client::Transaction, Error> {
+        let build_config = BuildConfig {
+            generate_abis: false,
+            generate_docs: true,
+            install_dir: self.move_options.output_dir.clone(),
+            ..Default::default()
+        };
+        let package = compile_move(build_config, self.move_options.package_dir.as_path())?;
```
[the remark this answers is not in the document]
Comment (gregnazario) [reply crates/aptos/src/move_tool/mod.rs:144]: Yeah, future PR

File: crates/aptos/src/move_tool/mod.rs [hunk 5]
```diff
@@ -98,3 +118,173 @@ fn compile_move(build_config: BuildConfig, package_dir: &Path) -> Result<Compile
...
+    #[clap(flatten)]
+    private_key_options: PrivateKeyInputOptions,
+    #[clap(flatten)]
+    move_options: MovePackageDir,
+    #[clap(flatten)]
+    node_options: NodeOptions,
+    #[clap(long)]
+    chain_id: ChainId,
+    #[clap(long, default_value_t = 1000)]
+    max_gas: u64,
```
[the remark this answers is not in the document]
Comment (gregnazario) [reply crates/aptos/src/move_tool/mod.rs]: We'll clean this up in the future, it's transparent to the users :)

File: crates/aptos/src/move_tool/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gregnazario) [reply crates/aptos/src/move_tool/mod.rs:147]: I've fixed this!

**Pull request closed without merging**